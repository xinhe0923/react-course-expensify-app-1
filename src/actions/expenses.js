import uuid from "uuid";
import database from "../firebase/firebase";
import expenses from "../tests/fixtures/expenses";
//component calls option operator
//action generator return object
//component dispatches object
//redux store changes

export const addExpense = (expense) => {
  return {
    type: "ADD_EXPENSE",
    expense,
  };
};

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = {
      description,
      note,
      amount,
      createdAt,
    };

    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then((ref) => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense,
          })
        ); //by return this we can toss on then to chain on in test file
      });
    //this function gets called internally be redux
    //gets called with dispatch
    //write some data to firebase, wait data to crrectly sync
    //use dispatch to add expense,
    //then the redux store reflect those changes
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id })); //by return this we can toss on then to chain on in test file
      });
  };
};

export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editExpense(id, updates)); //by return this we can toss on then to chain on in test file
      });
  };
};

//SET_
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses,
}); //change redux store

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;

    return database
      .ref(`users/${uid}/expenses`)
      .once("value")
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
// export const startSetExpenses;//fetch to firebase
