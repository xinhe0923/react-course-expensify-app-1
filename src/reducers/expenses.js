
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense]; //dont change original state
    case "REMOVE_EXPENSE":
      // console.log(action.id)

      return state.filter(({ id }) => id !== action.id);
    case "EDIT_EXPENSE":
      return state.map((expense) => {
        if (expense.id === action.id) {
          return { ...expense, ...action.updates };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};
