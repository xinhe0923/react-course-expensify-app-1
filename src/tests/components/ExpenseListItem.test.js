import  ExpenseListItem from '../../components/ExpenseListItem'
import React from 'react';
import {shallow} from 'enzyme';
import expenses from '../fixtures/expenses'


test('should render expenselist item with empty message',()=>{
    const wrapper=shallow(<ExpenseListItem {...expenses[0]} />)
    expect(wrapper).toMatchSnapshot();

})