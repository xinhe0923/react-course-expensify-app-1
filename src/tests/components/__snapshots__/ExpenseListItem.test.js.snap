// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenselist item with empty message 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1/"
  >
    <h3>
      gum
    </h3>
  </Link>
  <p>
    $1.95
    -
    January1 1st,1970
  </p>
</div>
`;
